packages/kbn-dev-utils/src/kbn_client/kbn_client_requester.ts

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Url from 'url';
+
+import Axios from 'axios';
+
+import { isAxiosRequestError, isConcliftOnGetError } from './errors';
+import { ToolingLog } from '../tooling_log';
+
+export const uriencode = (
+  strings: TemplateStringsArray,
+  ...values: Array<string | number | boolean>
+) => {
+  const queue = strings.slice();
+
+  if (queue.length === 0) {
+    throw new Error('how could strings passed to `uriencode` template tag be empty?');
+  }
+
+  if (queue.length !== values.length + 1) {
+    throw new Error('strings and values passed to `uriencode` template tag are unbalanced');
+  }
+
+  let url = queue.shift()!;
```

Comment:
Is this more straightforward?

```
return queue
  .map((val,i) => val + encodeURIComponent(values[i])
  .join('')
```

Reply:
Maybe, but there's one more item in the `queue` than the `values`, which is why the url starts with a single `queue.shift()`, so it'd need to be something like:

```
return queue.shift()! + queue.map((val, i) => encodeURIComponent(values[i] + val).join('')
```

I think I prefer the while loop personally, but I'll add a comment about the leading `.shift()`

Reply:
ended up refactoring this into a shift + reduce, not sure I think it's more straightforward (I love a plain and simple loop) but it didn't seem worse and felt in-line with what you were asking for.

Reply:
The loop with the comment is probably the clearest, but the reduce is fine too... I just didn't notice that `queue` had one more item when I originally looked at it